man/campagnol_rdv.8.in
.ds pidfilename @localstatedir@/run/campagnol_rdv.pid
.TH "CAMPAGNOL RDV SERVER" 8 "December 8, 2008" "@version@"
.SH NAME
campagnol_rdv - Decentralized VPN over UDP tunneling, RDV server
.SH SYNOPSIS
.B campagnol_rdv
.RB [ \-dghvV ]
.RB [ "-p \fIport\fR" ]
.RB [ "-m \fIn_clients\fR" ]
.SH DESCRIPTION
Campagnol is a distributed IP-based VPN software able to open new connections 
through NATs or firewalls without any configuration. It uses UDP for the
transport layer and utilizes tunneling and encryption (with DTLS).
.LP
This software is the rendezvous server of the VPN. It is used to register each
participant of the VPN and is needed for the UDP hole punching. It regularly
exchanges keepalive messages with each registered clients in order to maintain
the contexts of the connections in the involved NATs and firewalls.
.LP
The server opens a UDP socket to communicate with the clients. The default port
is \fI57888\fR.
.SH OPTIONS
.TP
.BR "\-d" ", " "--debug"
debug mode
.TP
.BR "\-D" ", " "--daemon"
start the client in background (daemonize) and create a PID file in
.I \*[pidfilename]
or in another file defined with
.I --pidfile
.TP
.BR "\-h" ", " "--help"
print help, then exit
.TP
.BR "\-m" ", " "--max-clients" "=\fI[maximum number of clients]\fR"
set the maximum number of registered clients. The default is 100 clients. Setting the value to 0 means unlimited (not recommended).
.TP
.BR "\-p" ", " "--port" "=\fI[port number]\fR"
set the server UDP port
.TP
.BR "\-P" ", " "--pidfile" "=\fIpidfile\fR"
write the pid into
.I pidfile
when running in background
.TP
.BR "\-v" ", " "--verbose"
verbose mode
.TP
.BR "\-V" ", " "--version"
show version information and exit
.SH SEE ALSO
.BR campagnol (8),
.BR campagnol.conf (5),
.SH AUTHORS
Florent Bondoux <fbondoux@users.sourceforge.net>


Changelog
v 0.3.3
	The client and rdv server have been ported to Cygwin.
	Add a texinfo manual.

	client:
	* bss_fifo: add a drop-tail mode in addition to the default blocking
	write.
	* communication, dtls_utils, peer: New design. Connections are now
	handled by a new thread from creation to shutdown. This is less tricky
	but cost the creation of a new thread as soon as a connection is
	asked.
	* configure.ac: Update the required version of OpenSSL to >=0.9.8j in
	order to have a decent DTLS implementation.
	* MTU: Allow the receiving of packets larger than the VPN MTU.
	* configure.ac: Add some optional warnings (with --enable-warnings)
	and a macro to check the CFLAGS.
	* communication: Fix some potential packet loss when the client
	re-registers with the RDV. The connection with the RDV server is now
	handled in a dedicated thread.
	* net_socket.c: Check the length of the interface name.
	* ca_wrap.sh: Add a command to list the generated certificates.
	* campagnol.c: Add a command line option to mlock the process.

v 0.3.2:
	client:
	Read the DTLS data (key and certificates) only once at startup.
	Add an option to give a directory with the root certificates (instead of
	giving a certificate chain file).
	Add the possibility to define which commands are run when the TUN
	device is opened or closed.
	The configuration files now accepts line continuation.
	Better CRL handling. Possibility to reload the CRL/key/certificate on
	signals.
	Correct the bad handling of timed-out connections or failed hole
	punching.
	Finer locking granularity.
	Linux: add options to set the TUN/TAP device TX queue.

	rdv server:
	Replace the java implementation by a new one in C.

v 0.3.1:
	client:
	Support rate limiting for the whole client and per connection.
	Give the PID file in argument instead of in the config file.
	OpenBSD port.
	Add an option to force the local address sent to the RDV (it may be
	usefull with multiple layers of NAT).
	Send keepalive messages on the connections. Add an option to control
	the interval between keepalive messages.
	Man page for the configuration file.

	rdv server:
	Add an option to set the maximum number of registered clients.
	Regularly clean dead clients and try to reduce memory consumption.
	Add a Makefile to ease building and installing.
	Man page.

v 0.3:
	client:
	Now using the autotools.
	FreeBSD port. Various cleanings in order to prepare the FreeBSD port and
	more robust code.
	New and faster SSL queues.
	Configuration: new INI-like format with a less buggy file parser. Now
	use a default configuration file (/etc/campagnol.conf).
	Better checking of loopback interfaces when searching the local
	IP and use getifaddrs when available.
	Fix some memory leaks with pthreads and OpenSSL.
	Write the PID into /var/run/campagnol.pid or another location.
	Add a receiving timeout so that openssl can detect a packet loss
	during a handshake and resend.
	Add the possibility to send the local IP and port to the RDV server.

	rdv server:
	Move the file into the net.sourceforge.campagnol package.
	Small cleanings.
	Better ANT build file.
	When two peers are behind the same NAT and their local addresses are
	kown, use the local adresses intead of the public addresses.

v 0.2:
	client: use syslog to log important messages when the client is
	started as a daemon. Add sample configuration files and certificates.
	A few bugfixes, most of them being unlikely to happen. Enhanced
	Makefile.
	rdv server: adjust the behavior of the verbose and debug parameters.

v 0.1:
	First release.


trunk/rdvserver/campagnol_rdv.c
/*
 * Campagnol RDV server, startup code
 *
 * Copyright (C) 2009 Florent Bondoux
 *
 * This file is part of Campagnol.
 *
 * Campagnol is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Campagnol is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Campagnol.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "rdv.h"

#include <getopt.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdint.h>

#include "../common/log.h"
#include "net_socket.h"
#include "server.h"

struct configuration config;
static char* pidfile = NULL;

volatile sig_atomic_t end_server = 0;

static void usage(void) {
    fprintf(stderr, "Usage: campagnol_rdv [OPTION]...\n\n");
    fprintf(stderr, "Options\n");
    fprintf(stderr, " -D, --daemon            fork in background\n");
    fprintf(stderr, " -d, --debug             debug mode. Can be use twice to dump the packets\n");
    fprintf(stderr, " -h, --help              this help message\n");
    fprintf(stderr, " -m, --max-clients=N     set the maximum number of connected clients\n");
    fprintf(stderr, " -P, --pidfile=FILE      write the pid into this file when running in background\n");
    fprintf(stderr, " -p, --port=PORT         listening port\n");
    fprintf(stderr, " -v, --verbose           verbose mode\n");
    fprintf(stderr, " -V, --version           show version information and exit\n\n");
    exit(EXIT_FAILURE);
}

static void version(void) {
    fprintf(stderr, "Campagnol VPN | Server | Version %s\n", VERSION);
    fprintf(stderr, "Copyright (c) 2007 Antoine Vianey\n");
    fprintf(stderr, "              2008-2009 Florent Bondoux\n");
    exit(EXIT_SUCCESS);
}

static void remove_pidfile(void) {
    unlink(pidfile);
    free(pidfile);
}

static void create_pidfile(void) {
    int fd;
    FILE *file;

    if (strlen(pidfile) == 0)
        return;

    if (unlink(pidfile) != 0 && errno != ENOENT) {
        return;
    }
    fd = open(pidfile, O_CREAT|O_WRONLY|O_TRUNC|O_NOFOLLOW, (mode_t) 00644);
    if (fd == -1) {
        return;
    }
    file = fdopen(fd, "w");
    if (file == NULL) {
        return;
    }

    fprintf(file, "%d\n", getpid());

    fclose(file);

    atexit(remove_pidfile);
}

static void daemonize(void) {
    int r;

    printf("Going in background...\n");
    r = daemon(1, 0);
    if (r != 0) {
        log_error(errno, "Unable to daemonize");
        exit(EXIT_FAILURE);
    }

    config.verbose = 0;
    config.debug = 0;
}

static void sig_handler(int sig) {
    switch (sig) {
        case SIGINT:
        case SIGTERM:
        case SIGQUIT:
            end_server = 1;
            log_message("received signal %d, exiting...", sig);
            break;
        default:
            break;
    }
}

static int parse_args(int argc, char **argv) {
    int opt;
    int i;
    char *endptr;

    config.verbose = 0;
    config.daemonize = 0;
    config.debug = 0;
    config.dump = 0;
    config.serverport = SERVER_PORT_DEFAULT;
    config.max_clients = MAX_CLIENTS_DEFAULT;
    config.pidfile = NULL;

    struct option long_options[] = {
        {"daemon", 0, NULL, 'D'},
        {"debug", 0, NULL, 'd'},
        {"help", 0, NULL, 'h'},
        {"max-clients", 1, NULL, 'm'},
        {"pidfile", 1, NULL, 'P'},
        {"port", 1, NULL, 'p'},
        {"verbose", 0, NULL, 'v'},
        {"version", 0, NULL, 'V'},
        {0, 0, 0, 0}
    };
    while ((opt = getopt_long(argc, argv, "dDhm:p:P:vV", long_options, NULL)) >= 0) {
        switch(opt) {
            case 'D' :
                config.daemonize = 1;
                break;
            case 'd' :
                if (config.debug) {
                    config.dump = 1;
                }
                else {
                    config.debug = 1;
                    config.verbose = 1;
                }
                break;
            case 'h' :
                return 1;
            case 'm':
                errno = 0;
                config.max_clients = (int) strtol(optarg, &endptr, 10);
                if (errno != 0 || endptr == optarg || config.max_clients < 0) {
                    return 1;
                }
                break;
            case 'p' :
                i = atoi(optarg);
                if (i <= 0 || i > UINT16_MAX) {
                    return 1;
                }
                config.serverport = (uint16_t) i;
                break;
            case 'P' :
                config.pidfile = CHECK_ALLOC_FATAL(strdup(optarg));
                break;
            case 'v':
                config.verbose = 1;
                break;
            case 'V' :
                return 2;

            default : return 1;
        }
    }

    argv += optind;
    argc -= optind;

    if (argc != 0) {
        return 1;
    }

    return 0;
}

int main(int argc, char **argv) {
    int pa;
    int sockfd;
    int exit_status = EXIT_SUCCESS;

    pa = parse_args(argc, argv);
    if (pa == 1) {
        usage();
    }
    else if (pa == 2) {
        version();
    }

    if (config.daemonize) daemonize();
    log_init(config.daemonize, config.verbose, "campagnol_rdv");

    if (config.daemonize) {
        const char *pidtmp = (config.pidfile != NULL) ? config.pidfile : DEFAULT_PID_FILE;
        pidfile = CHECK_ALLOC_FATAL(strdup(pidtmp));
        create_pidfile();
    }

    sockfd = create_socket();
    if (sockfd < 0) {
        exit_status = EXIT_FAILURE;
        goto clean_end;
    }

    signal(SIGINT, sig_handler);
    signal(SIGQUIT, sig_handler);
    signal(SIGTERM, sig_handler);

    rdv_server(sockfd);

    clean_end:

    if (sockfd > 0)
        close(sockfd);

    log_close();

    if (config.pidfile)
        free(config.pidfile);

    exit(exit_status);
}
